vivisect/impemu/emulator.py

```diff
@@ -88,6 +73,56 @@ def __init__(self, vw, logwrite=False, logread=False):
 
             self.hooks[impname] = val
 
+        self.stack_map_mask = None
+        self.stack_map_base = None
+        self.stack_map_top = None
+        self.stack_pointer = None
+        self.initStackMemory()
+
+    def initStackMemory(self, stacksize=4096):
+        '''
+        Setup and initialize stack memory.
+        You may call this prior to emulating instructions.
+        '''
+        if self.stack_map_base is None:
+            self.stack_map_mask = e_bits.sign_extend(0xfff00000, 4, self.vw.psize)
+            self.stack_map_base = e_bits.sign_extend(0xbfb00000, 4, self.vw.psize)
+            self.stack_map_top = self.stack_map_base + stacksize
+            self.stack_pointer = self.stack_map_top
+
+            # Map in a memory map for the stack
+            stack_map = init_stack_map
+            if stacksize != 4096:
+                stack_map = ''.join([struct.pack('<I', self.stack_map_base+(i*4))
+                                        for i in xrange(stacksize)])
+
+            self.addMemoryMap(self.stack_map_base, 6, "[stack]", stack_map)
+            self.setStackCounter(self.stack_pointer)
+
+            # Create some pre-made taints for positive stack indexes
+            # NOTE: This is *ugly* for speed....
+            taints = [ self.setVivTaint('funcstack', i * self.psize) for i in xrange(20) ]
+            taintbytes = ''.join([ e_bits.buildbytes(taint,self.psize) for taint in taints ])
+
+            self.writeMemory(self.stack_pointer, taintbytes)
+        else:
+            existing_map_size = self.stack_map_top - self.stack_map_base
+            new_map_size = stacksize - existing_map_size
+            if new_map_size < 0:
+                raise RuntimeError('cannot shrink stack')
```

Comment:
this may not be appropriate (at least, maybe use a custom exception type). have any suggestions?